pkg/client/client.go

```diff
@@ -234,6 +243,80 @@ func (o *OptimizelyClient) decideAll(userContext OptimizelyUserContext, options
 	return o.decideForKeys(userContext, allFlagKeys, options)
 }
 
+func (o *OptimizelyClient) fetchQualifiedSegments(userContext *OptimizelyUserContext, options []pkgOdpSegment.OptimizelySegmentOption, callback func(segments []string, err error)) {
+	var err error
+	defer func() {
+		if r := recover(); r != nil {
+			switch t := r.(type) {
+			case error:
+				err = t
+			case string:
+				err = errors.New(t)
+			default:
+				err = errors.New("unexpected error")
+			}
+			o.logger.Error("fetchQualifiedSegments call, optimizely SDK is panicking with the error:", err)
+			o.logger.Debug(string(debug.Stack()))
+		}
+	}()
+
+	// on failure, qualifiedSegments should be reset if a previous value exists.
+	userContext.SetQualifiedSegments(nil)
+
+	_, err = o.getProjectConfig()
+	if err != nil {
+		if callback != nil {
+			callback(nil, decide.GetDecideError(decide.SDKNotReady))
+		}
+		return
+	}
+
+	fetchAndSetSegments := func() (segments []string, err error) {
+		qualifiedSegments, segmentsError := o.OdpManager.FetchQualifiedSegments(userContext.GetUserID(), options)
+		if segmentsError == nil {
+			userContext.SetQualifiedSegments(qualifiedSegments)
+		}
+		return qualifiedSegments, segmentsError
+	}
+
+	// Async
+	if callback != nil {
+		go func() {
+			callback(fetchAndSetSegments())
```

Comment:
there may be inconsistency / race condition issue when you call routine. thoughts?

Reply:
All internal implementations are using locks where required so this should not be a problem.